module/applications/journal/spell-list-page-sheet.mjs

```diff
@@ -0,0 +1,177 @@
+    return options;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Different ways in which spells can be grouped on the sheet.
+   * @type {Record<string, string>}
+   */
+  static get GROUPING_MODES() {
+    return SpellListJournalPageData.GROUPING_MODES;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Currently selected grouping mode.
+   * @type {string|null}
+   */
+  grouping = null;
+
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  get template() {
+    return `systems/dnd5e/templates/journal/page-spell-list-${this.isEditable ? "edit" : "view"}.hbs`;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  async getData(options) {
+    const context = super.getData(options);
+    context.CONFIG = CONFIG.DND5E;
+    context.system = context.document.system;
+
+    context.title = Object.fromEntries(Array.fromRange(4, 1).map(n => [`level${n}`, context.data.title.level + n - 1]));
+
+    context.description = await TextEditor.enrichHTML(context.system.description.value, {
+      async: true, relativeTo: this
+    });
+    if ( context.description === "<p></p>" ) context.description = "";
+
+    context.GROUPING_MODES = this.constructor.GROUPING_MODES;
+    context.grouping = this.grouping || context.system.grouping;
+
+    context.spells = await this.prepareSpells(context.grouping);
+
+    context.sections = {};
+    for ( const spell of context.spells ) {
+      let section;
+      switch ( context.grouping ) {
+        case "level":
+          const level = spell.system.level;
+          section = context.sections[level] ??= { header: CONFIG.DND5E.spellLevels[level], spells: [] };
+          break;
+        case "school":
+          const school = spell.system.school;
+          section = context.sections[school] ??= { header: CONFIG.DND5E.spellSchools[school]?.label, spells: [] };
+          break;
+        case "alphabetical":
+          const letter = spell.name.slice(0, 1).toLowerCase();
+          section = context.sections[letter] ??= { header: letter.toUpperCase(), spells: [] };
+          break;
+        default:
+          continue;
+      }
+      section.spells.push(spell);
+    }
+    if ( context.grouping === "school" ) context.sections = sortObjectEntries(context.sections, "header");
+
+    return context;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Load indices with necessary information for spells.
+   * @param {string} grouping  Grouping mode to respect.
+   * @returns {object[]}
+   */
+  async prepareSpells(grouping) {
+    let fields;
+    switch ( grouping ) {
+      case "level": fields = ["system.level"]; break;
+      case "school": fields = ["system.school"]; break;
+      default: fields = []; break;
+    }
+
+    let collections = new Collection();
+    for ( const uuid of this.document.system.spells ) {
+      const { collection } = foundry.utils.parseUuid(uuid);
```

Comment:
A non-compendium Item has a `collection`, but you can't call `collection.getIndex` on it, so it throws an error and prevents the sheet from rendering. It's probably not a good idea for a user to use spells that are not in a compendium, but we could fix the error and allow it. Alternatively we could check in `_onDrop` and emit a warning, preventing the Item from being added if it's not in a compendium.